File: implementation using Matlab/convertHtoG.m
function [G, k] = convertHtoG(H)
% this function is to convert a general parity-check matrix H into a
% corresponding generator matrix G
% input : a general parity-check matrix H
% output : the corresponding generator matrix G, the dimension k

H_ = H_sf(H);  

current_row = 0;
[num_row, num_col] = size(H);
n = num_col;

% removing all the all-zero rows in H_
for row = 1 : num_row
    if (sum(H_(row,:)) ~= 0)
        current_row = current_row + 1;
        full_H(current_row,:) = H_(row,:);
    end
end

total_row = current_row; 

% the rank of parity-check matrix H
R_H = total_row;

% select all the independent columns from full-row-rank full_H
% current_col = 1;
% current_row = 1;
% Q_num = 1;
independent_cols =zeros(1,total_row);
% Q_cols = zeros(1, n-R_H);
inv_permutation = zeros(1, n); 

% for num = R_H + 1 : n
%     Q_cols(num-R_H) = num;
% %     inv_permutation(num - R_H) = num;
%     inv_permutation(num) = num - R_H;
% end

Q_cols = 1 : n;
for num = 1 : R_H
    if full_H(num, num) == 1
        independent_cols(num) = num;
        inv_permutation(num) = num + n - R_H;
        location = find(Q_cols == num);
        Q_cols(location) = [];   %delete this independent column from Q_cols
    else 
        location = 0;
        for col = Q_cols
            if full_H(num, col) == 1
                independent_cols(num) = col;
                inv_permutation(col) = num + n -R_H;
                location = find(Q_cols == col);
                break;
            end
        end
        if location ~= 0
            Q_cols(location) = [];
        end
    end
end

col = 0;
for num = Q_cols
    col = col +1;
    inv_permutation(num) = col;
end
        
% while (current_row <= total_row) 
%     
%     if full_H(current_row, current_col) == 1
%         independent_cols(current_row) = current_col ;
%         inv_permutation(current_col) = current_row + n -R_H;
%         current_row = current_row +1;
%         current_col = current_col +1;
%     else
%         Q_cols(Q_num) = current_col;
%         inv_permutation(current_col) = Q_num;
%         Q_num = Q_num + 1;
%         current_col = current_col +1;
%     end
% end

k = n - R_H;  % dimension of our code;


% construct the permutated generator matrix G_
G_ = eye(k);
for num = 1 : k
    G_(num, k+1:n) = full_H(:,Q_cols(num))';
end

% using the inverse permutation to recover the generator matrix
% corresponding to the original parity-check matrix,i.e. the original code;

for num = 1 : n
    G(:,num) = G_(:, inv_permutation(num));
end    

File: implementation using C/LDPC-related Lib/ConvertHtoG.c
#include "ConvertHtoG.h"
#include "Htrsf.h"
#include <stdlib.h>
#include <stdio.h>

void ConvertHtoG(char *H, int n, int m, char **G, int *dim)
{
    char *full_H = (char *)calloc(n*m, sizeof(char));
    int *indpdt_cols = NULL;
    int *other_cols = NULL;
    int *inv_permutation = (int *)calloc(m, sizeof(int));
    int current_row = 0, sum = 0, k = 0;
    int flag = 0, num = 0, i = 0;
    int row = 0, col = 0, Rank_H = 0;
    char *Gt = NULL;


    // firstly, removing all the all-zero rows in the parity-check matrix H to get full-rank full_H
    for (row = 0; row < n; row++)
    {
        sum = 0;
        for (col = 0; col < m; col++)
            sum += *(H+(row*m+col));
        if (sum != 0)
        {
            for (col = 0; col < m; col++)
            {
                *(full_H + (current_row*m + col)) = *(H + (row*m + col));
            }
            current_row ++;
        }
    }

    Rank_H = current_row;  //the Rank of H is actually equivalent to the number of rows in full_H;
    
    indpdt_cols = (int *)calloc(Rank_H, sizeof(int));

    for (num=0; num < Rank_H; num++)
        if (*(full_H+(num*m+num)) == 1)
        {
  //          printf("num=========%d\n", num);
  //              getchar();
            *(indpdt_cols + num) = num;
            *(inv_permutation + num) = num + m - Rank_H;

        }
        else
        {
            for (col=0; col < m; col++)
            {
                flag = 1;
                for (i = 0; i < num; i++)
                    if (col == *(indpdt_cols + i))
                    {
                        flag = 0;
                        break;
                    }
                if (flag == 0)
                    continue;

                if  (*(full_H + (num*m+col)) == 1)
                {
                    *(indpdt_cols + num) = col;
                    *(inv_permutation + col) = num + m - Rank_H;
                    break;
                }
            }
        }

    num = 0;
    other_cols = (int *)calloc(m - Rank_H, sizeof(int));

    for (col = 0; col <m; col++)
    {
        flag = 1;
        for (i = 0; i < Rank_H; i++)
            if ( col == *(indpdt_cols + i))
            {
                flag = 0;
                break;
            }
        if (flag)
        {
            *(inv_permutation + col) = num;
            *(other_cols + num) = col;
            num ++;
        }

    }


//    for (i = 0; i < Rank_H; i++)
//        printf("i====%d, indpdt_col = %d\n", i, *(indpdt_cols + i));


   
    k = m - Rank_H;

    Gt =  (char *)calloc(k*m, sizeof(char));


    // Construct the permutated generator matrix Gt
    for (row = 0; row < k; row++)
    {
        *(Gt + (row*m + row)) = 1;
        for (num = 0; num < Rank_H; num++)
            *(Gt + (row*m + k + num)) = *(full_H + (num*m + *(other_cols + row)));
    }

    *G =  (char *)calloc(k*m, sizeof(char));
      
    for (col = 0; col < m; col++)
        for (row = 0; row < k; row++)
        {
            *((*G) + (row * m + col)) = *(Gt + (row * m + (*(inv_permutation + col))));
     //       printf("Gt----------%d\n", row * m + *(inv_permutation + col));
        }
   
   
    *dim = k;


    free(indpdt_cols);
    free(other_cols);
    free(inv_permutation);
    free(full_H);
    free(Gt);
    return;
}





 